### PR TITLE
Support for removing functions specified by the user

### PR DESCRIPTION
**Pull request by shoaibCS**
support for removing functions specified by the user

### CODE REVIEW

File: razor/passes.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def interface(input_file, output_file, wrt, use_seadsa):
     args = ['-Pinterface']
     if use_seadsa:
         args += ['-Pinterface-with-seadsa',
-                 # improve precision of sea-dsa by considering types            
+                 # improve precision of sea-dsa by considering types
```
Comment (caballa) [inline razor/passes.py:58]: Similar to Rafae's PR. Your editor is adding extra white spaces that make hard to review the PR

File: src/transforms/RemoveFunctions.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+using namespace std;
+
+struct RemoveFunctions : public FunctionPass {
+	
+	static char ID;
+	RemoveFunctions() : FunctionPass(ID) {}
+        
+	bool runOnFunction(Function &F) override {
+
+		Module* module = F.getParent();
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:50]: use LLVM `cl::list<string>` for defining an user option that consists of a list of strings. (https://llvm.org/docs/CommandLine.html). In that way, you don't need to assume one single string that actually contains multiple strings separated by comma and then do the parsing by yourself

File: src/transforms/RemoveFunctions.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		assert(printfFunc && "printf not found in module");
+
+		// Declare exit function
+		Type *voidType = Type::getVoidTy(module->getContext());
+		vector<Type *> exitArgsTypes;
+		exitArgsTypes.push_back(Type::getInt32Ty(module->getContext()));
+		FunctionType *exitType = FunctionType::get(voidType, exitArgsTypes, false);
+		Function *exitFunc = dyn_cast<Function>(module->getOrInsertFunction("exit", exitType).getCallee());
+
+		string functionName = F.getName().str();
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:80]: you can print directly `F.getName()` that returns a StringRef which much more preferable than std::string

File: src/transforms/RemoveFunctions.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+			//Add printf call in empty BB
+            		Value *str = Builder.CreateGlobalStringPtr("You asked to remove this function. Program terminating!\n", "str");
+            		std::vector<Value *> removeMessage({str});
+            		Builder.CreateCall(printfFunc, removeMessage, "unreahcable");
+           		//Add exit call in empty BB
+            		Value *status = ConstantInt::get(Type::getInt32Ty(module->getContext()),(-10));
+            		Builder.CreateCall(exitFunc, status, "");
+            		Type* returnType = F.getReturnType();
+            		//Add terminator instruction
+            		ReturnInst::Create(module->getContext(), UndefValue::get(returnType), emptyBB);
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:99]: Add a LLVM `unreachable` instruction. Return something with undef value is not a good idea.

File: src/transforms/RemoveFunctions.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+		while (string_to_stream.good()) {
+			string substr;
+        		getline(string_to_stream, substr, ',');
+        		removeFunctions.push_back(substr);
+    		}
+
+//		vector<string> removeFunctions{"func1","func2","func3","func4","add"};
+    
+		// Declare printf function
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:62]: Create a function `callPrintf` or something to help reading the code

File: src/transforms/RemoveFunctions.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+//		std::vector<Type *> printfArgsTypes({Type::getInt8PtrTy(module->getContext())});
+//		FunctionType *printfType = FunctionType::get(intType, printfArgsTypes, true);
+//		auto *printfFunc = dyn_cast<Function>(module->getOrInsertFunction("printf", printfType).getCallee()); 
+		Type* charType = Type::getInt8PtrTy(module->getContext());
+                FunctionType *printf_type = FunctionType::get(charType,true);
+                auto *printfFunc = cast<Function>(module->getOrInsertFunction("printf", printf_type).getCallee());
+
+		assert(printfFunc && "printf not found in module");
+
+		// Declare exit function
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:73]: Create a function `callExit`

File: src/transforms/RemoveFunctions.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		Type *voidType = Type::getVoidTy(module->getContext());
+		vector<Type *> exitArgsTypes;
+		exitArgsTypes.push_back(Type::getInt32Ty(module->getContext()));
+		FunctionType *exitType = FunctionType::get(voidType, exitArgsTypes, false);
+		Function *exitFunc = dyn_cast<Function>(module->getOrInsertFunction("exit", exitType).getCallee());
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:78]: Not sure if this will call to the C `exit` function

File: src/transforms/RemoveFunctions.cpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+//		FunctionType *printfType = FunctionType::get(intType, printfArgsTypes, true);
+//		auto *printfFunc = dyn_cast<Function>(module->getOrInsertFunction("printf", printfType).getCallee()); 
+		Type* charType = Type::getInt8PtrTy(module->getContext());
+                FunctionType *printf_type = FunctionType::get(charType,true);
+                auto *printfFunc = cast<Function>(module->getOrInsertFunction("printf", printf_type).getCallee());
+
+		assert(printfFunc && "printf not found in module");
+
+		// Declare exit function
+		Type *voidType = Type::getVoidTy(module->getContext());
```
Comment (caballa) [inline src/transforms/RemoveFunctions.cpp:74]: Remove all the basic blocks and instructions from the function

**Comment by caballa** [comment]
Please create a different PR for trimmer.set

**Comment by caballa** [comment]
The last two commits c0fcbba and 63bcde2 should be squashed together

File: src/transforms/RemoveFunctions.cpp (hunk 4 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp]: Done. I

File: src/transforms/RemoveFunctions.cpp (hunk 5 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp]: Done. I created `makePrintf` to follow the convention Rafae was following.

File: src/transforms/RemoveFunctions.cpp (hunk 6 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp]: Done.

File: src/transforms/RemoveFunctions.cpp (hunk 8 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp:57]: Done.

File: src/transforms/RemoveFunctions.cpp (hunk 3 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp]: Done. Replaced string with StringRef.

File: src/transforms/RemoveFunctions.cpp (hunk 2 above)
Comment (shoaibCS) [reply src/transforms/RemoveFunctions.cpp]: Done.

**Comment by caballa** [comment]
@shoaibCS : what is the status of this PR? You need to resolve some conflics

**Comment by shoaibCS** [comment]
> @shoaibCS : what is the status of this PR? You need to resolve some conflicts

@caballa : I have resolved the two conflicts. can you kindly tell if everything is fine and then accept the pull request? Thanks!

**Pull request merged**